etc/calc_release_version.py

```diff
@@ -152,77 +180,104 @@ def get_next_minor(prerelease_marker):
+        tag_glob = 'r'+  release_branch_match.group('vermaj') + '.' \
                 + release_branch_match.group('vermin') + '.*'
-        tags = check_output(['git', 'tag', '--list', tag_glob])
-    else:
-        # Not a release branch, so look for the most recent tag in history
-        commits = check_output(['git', 'log', '--pretty=format:%H',
-                                '--no-merges'])
-        if len(commits) > 0:
-            for commit in commits.splitlines():
-                tags = check_output(['git', 'tag', '--points-at',
-                                     commit, '--list', 'r*'])
-                if len(tags) > 0:
-                    # found a tag, we should be done
-                    break
-
-    return tags
-
-def process_and_sort_tags(tags):
+        got = check_output(['git', 'tag', '--list', tag_glob]).splitlines()
+        debug('Got tags: {0}'.format(got))
+        return got
+
+    # Not a release branch, so look for the most recent tag in history
+    commits = check_output(['git', 'log', '--pretty=format:%H', '--no-merges'])
+    tags_by_obj = get_object_tags()
+    for commit in commits.splitlines():
+        got = tags_by_obj.get(commit)
+        if got:
+            debug('Got tags: {0}'.format(got))
+            return got
+    # No tags
+    debug('(No tags found for this branch)')
+    return []
+
+
+def iter_tag_lines():
+    """
+    Generate a list of pairs of strings, where the first is a commit hash, and
+    the second is a tag that is associated with that commit. Duplicate commits
+    are possible.
+    """
+    output = check_output(['git', 'tag', '--list', '--format=%(*objectname)|%(objectname)|%(refname:strip=2)'])
+    lines = output.splitlines()
+    for l in lines:
+        obj, tagobj, tag = l.split('|', 2)
+        if re.match(r'r\d+\.\d+', tag):
+            yield obj, tagobj, tag
+
+
+def get_object_tags():  # type: () -> dict[str, list[str]]
+    """
+    Obtain a mapping between commit hashes and a list of tags that point to
+    that commit. Untagged commits will not be included in the resulting map.
+    """
+    ret = {}  # type: dict[str, list[str]]
+    for obj, tagobj, tag in iter_tag_lines():
+        if obj:
+            ret.setdefault(obj, []).append(tag)
+        ret.setdefault(tagobj, []).append(tag)
+    return ret
+
+
+def process_and_sort_tags(tags):  # type: (list[str]) -> list[str]
     """
     Given a string (as returned from get_branch_tags), return a sorted list of
     zero or more tags (sorted based on the Version comparison) which meet
     the following criteria:
-        - a final release tag (i.e., r3.x.y without any pre-release suffix)
+        - a final release tag (i.e., 1.x.y without any pre-release suffix)
         - a pre-release tag which is not superseded by a release tag (i.e.,
-          r3.x.y-preX iff r3.x.y does not already exist)
+          1.x.y-preX iff 1.x.y does not already exist)
```

Comment:
```suggestion
          r3.x.y-preX iff r3.x.y does not already exist)
```
It would be good to preserve the C++ driver-style tag names in the comments

Reply:
I copy-pasted the script over from C driver and had to fix it up to work with the C++ style branches and tags, so some things were missed.

---

Aside: In `iter_tag_lines`, I discovered that some C++ repo tags are pointing to the commits directly, while others are pointing to separate "tag objects." This is a Git "feature" (quirk?) I was unaware of. The C driver uses exclusively discrete tag objects rather than tags glued directly to commits, so it took me took a while to discover why it wasn't working in this repository. I have no idea how/when this distinction occurs. It appears that only the `debian/*` tags have separate tag objects in the C++ repo.